Simplify timelines UI/reuse UI elements

tillprochaska opened the pull request:
See https://github.com/alephdata/aleph/issues/2673 for context.

## Scope of this PR

The test cases in the top-level test suite give a good impression of the main features:

```
PASS src/components/Timeline2/Timeline.test.tsx (7.005 s)
  ✓ sorts and filters items by temporal start (115 ms)
  ✓ allows creating new items (2015 ms)
  ✓ does not allow creating new items if not writeable (35 ms)
  ✓ allows removing entities (49 ms)
  ✓ does not allow removing entities if not writeable (49 ms)
  Sidebar
    ✓ allows changing entity color (120 ms)
    ✓ does not allow changing entity color if not writeable (35 ms)
    ✓ allows changing entity properties (190 ms)
    ✓ does not allow changing entity properties if not writeable (47 ms)
    ✓ allows changing entity-type properties (335 ms)
  Empty state
    ✓ displays a message (8 ms)
    ✓ allows creating new items (45 ms)
    ✓ does not allow creating new items if not writeable (2 ms)
```

Timelines will eventually have two different views (or renderers), a list view and a chart view. This PR mainly deals with refreshing the list view. However, I have already implemented a first version of the chart view (in `src/components/Timeline2/TimelineChart`). The chart view is not exposed in the UI right now.

In an ideal world, the chart view should be in a different PR. That would’ve resulted in more work (or splitting out this part into a separate PR and potentially losing part of the Git history).

I’d suggest that you either ignore the `TimelineChart` directory when reviewing this PR (instead reviewing it later, when we actually enable this view in the UI) or that you review what’s there now already. If you decide to do the latter, you can manually enable the chart view by adding a `renderer="chart"` property [here](https://github.com/alephdata/aleph/blob/till/timelines/ui/src/screens/TimelineScreen/TimelineScreen.jsx#L195-L204).

## How to review

I’d recommend reviewing in the following order:

1. Take a look at the types in `src/components/Timeline2/types.ts` and the classes (`TimelineItem`, `ImpreciseDate`) in `src/components/Timeline2/util.ts`.
2. `src/screens/TimelineScreen/TimelineScreen.jsx`
3. `src/components/Timeline2/Timeline.tsx` and `src/components/Timeline2/Timeline.test.tsx`
4. Then take a look at any components referenced in the `Timeline` component as well as respective tests.
5. Anything else

I try to keep my git history clean and often add relevant implementation considerations to the commit message body. In case you’re wondering about something specific, it may make sense to click "View file" on GitHub and then "Blame" to find commits that have changed a specific file.

Also read through the implementation notes below.

## Implementation notes

While this PR replaces most of the existing timelines implementation, I have kept the old implementation around while developing and have selectively used bits and pieces of the existing implementation. That is why there is a `src/components/Timeline` directory with the old implementation and a `src/components/Timeline2` directory with the new implementation.

I will remove the old implementation before we merge this PR and rename the `Timeline2` directory. This will remove about 1500 LOC, so the net-lines added in this PR will be closer to ~2500 rather than the ~4000 this PR has at the moment. Keep in mind that a big part of the lines added is for test coverage (this feature wasn’t tested before).

Apart from that, I have partially re-implemented the existing entity sidebar (that is already used in network diagrams) with the goal of porting it to TypeScript and refactoring it to make it easier to test. Right now, there exist two implementations of the same UI:

* Old implementation used for network diagrams in `src/react-ftm/components/NetworkDiagram/toolbox/EntityViewer.tsx`
* New implementation used in timelines in `src/components/Timeline2/EntityViewer2.tsx`

The mid-term goal is to replace the old implementation used in network diagrams. In order to keep the scope of this PR smaller, and to keep the risk of implementing bugs in network diagrams low, I haven’t done this in this PR.

## To do

**Before review:**
- [x] Remove entities from timeline
- [x] Fix entity suggestions in entity viewer sidebar
- [x] Fix edge selects in create dialog
- [x] Improve keyboard navigation
  - [x] Up/down arrows to select prev/next item
  - [x] Clicking an item should set keyboard focus to it
  - [x] Pressing escape should close entity sidebar
- [x] Use table-like layout (two columns for dates)
- [ ] ~Rebase and reenable type checking in tests where we use window.crypto~
- [x] Fix bug: Add entity and delete directly fails
- [x] Add proper translations
- [x] Add read-only mode (depending on permissions)

**Before merging:**
- [x] Delete `Timeline` directory and rename `Timeline2` directory
- [x] Remove unrelated changes (see review comment)
- [x] Remove script to populate timelines during development

**Potential enhancements:**
- [x] Add status indicator ("Saved"/"Saving…"/"Error"/"Offline")
- [ ] Display items without temporal extent separately
- [x] Improve ranking/order of properties in entity viewer, always display temporal properties first?
- [x] Add empty state
- [x] Add loading state
- [ ] Format dates
- [x] Add tooltip to remove icon button
- [ ] Use `TimelineItem` in `EntityViewer`
- [ ] Don’t let people add items without at least one date field filled out
- [ ] Move "Add item" button to the fixed entity set bar?
- [ ] Add indicator for optional fields when creating a timeline item
- [ ] Sort list by start or end date (asc or desc)

## Screenshots

![Screen Shot 2023-01-11 at 16 49 56](https://user-images.githubusercontent.com/1512805/211852131-8830ec4f-9eed-486b-b018-6bce8377556f.png)

![Screen Shot 2023-01-11 at 16 48 29](https://user-images.githubusercontent.com/1512805/211852155-e2c3cff8-dc0b-4944-bdce-42e789cf317d.png)

![Screen Shot 2023-01-11 at 16 52 22](https://user-images.githubusercontent.com/1512805/211852762-05326388-45b4-4b85-8263-3a20275dc563.png)

tillprochaska commented on ui/src/components/Toolbar/DialogToggleButton.jsx:
```diff
@@ -32,6 +32,8 @@ class DialogToggleButton extends React.Component {
         <Dialog
           isOpen={this.state.isOpen}
           toggleDialog={this.toggleDialog}
+          onMouseOver={(event) => event.stopPropagation()}
+          onMouseOut={(event) => event.stopPropagation()}
           {...dialogProps}
```
This change shouldn’t be in here. Looks like a solution to #2684 which I accidentally included in a commit on this branch.

tillprochaska commented on ui/src/components/Timeline2/util.ts:
```diff
@@ -0,0 +1,212 @@
+  items: Array<TimelineItem>,
+  onUnselect: () => void
+) {
+  const itemRefs = useMemo(() => items.map(() => createRef<T>()), [items]);
```
```suggestion
  const itemRefs = useMemo(() => items.map(() => useRef<T>()), [items]);
```

Rosencrantz commented on ui/src/components/Timeline2/Timeline.tsx:
```diff
@@ -0,0 +1,147 @@
+
+  const createButton = (
+    <Button intent={Intent.PRIMARY} icon="add" onClick={toggleCreateDialog}>
+      <FormattedMessage id="timeline.add_item" defaultMessage="Add item" />
```
I thought you added code for i18n here?

tillprochaska replied:
Yes, the `FormattedMessage` component is a component exposed by `react-intl` it will render a translation if available for the current language or the default message:

https://formatjs.io/docs/react-intl/components#formattedmessage

Rosencrantz reviewed (approved):
Not a massive fan of the Timeline2, but that is just me, and by no means a request for change or a need to refactor later. 

Other than that, this is a big PR so I've commented where I found myself puzzled and took the time to run through the UI locally.

tillprochaska commented on ui/src/app/utils.scss:
```diff
@@ -0,0 +1,11 @@
+  overflow: hidden;
+  padding: 0;
+  position: absolute;
+  width: 1px;
```
Yes, this is a utility class to hide elements visually while still keeping it accessible to screen readers: https://css-tricks.com/inclusively-hidden/

tillprochaska commented on ui/src/components/Timeline2/util.ts:
```diff
@@ -0,0 +1,257 @@
+  const [suggestions, setSuggestions] = useState<Array<Entity>>([]);
+  const [isFetching, setIsFetching] = useState(false);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
When using the `useCallback` hook in React, we need to specify the dependencies (i.e. any variables or functions used in this hook) as a second argument. In case of `useCallback`, React guarantees that it will reuse the same function on rerenders as long as the dependencies do not change.

Not specifying all dependencies of the `useCallback` hook can lead to bugs, because the callback function may not be updated (but we’d expect it to update). That’s why we have a linter rule that checks that we always specify all dependencies when using `useCallback` (and other similar hooks).

So far so good. Our usage is a little special here, because we also use `throttle` from lodash, a helper that ensures that we use to ensure that we do not fetch suggestions more frequently than once per 200ms. The linter doesn’t know how exactly throttle works and what it returns, so it can’t be sure that this doesn’t violate the rule to always specify all hook dependencies.

We do however know that `throttle` is deterministic and the returned function will always behave in the exact same way as long as we pass it the same arguments. So I’ve ignored this specific linter rule in this specific case.

tillprochaska commented on ui/src/components/Timeline2/TimelineList/TimelineListItem.scss:
```diff
@@ -0,0 +1,83 @@
+}
+
+.TimelineListItem > td:first-child {
+  border-left: 4px solid var(--timeline-item-color);
```
Yes, this is a decorative border, so I think using a px value here is fine.

tillprochaska commented on ui/src/components/Timeline2/Timeline.scss:
```diff
@@ -0,0 +1,48 @@
+@import 'app/variables.scss';
+
+.Timeline {
+  --viewer-width: 275px;
```
True, this follows the existing sidebar styles in network diagrams which uses this value. Should probably add a comment for everyone else who doesn’t have that background.

tillprochaska commented on ui/src/components/Timeline2/util.ts:
```diff
@@ -0,0 +1,257 @@
+  //
+  // Setting the day to zero will return the last day of the previous month:
+  // new Date(2000, 1, 0) // Jan 31st
+  // new Date(2000, 12, 0) // Dec 31st
```
True, this is because I’ve used the `/** ... */` style for JSDoc-like class comments, whereas `// ...` is used for comments describing implementation details. But I can change that to use `/* ... */` for those if you prefer.

tillprochaska commented on ui/src/components/Timeline2/TimelineList/TimelineListItem.scss:
```diff
@@ -0,0 +1,83 @@
+
+// Double class is required in order to increase specificity over the
+// default Blueprint table styles.
+.TimelineListItem__actions.TimelineListItem__actions {
```
This workaround is required to increase the specificity as otherwise the styles would be overridden by default Blueprint styles. Chaining the same class name twice keeps the selector as low as possible while increasing it as much as necessary.

[This section](https://cssguidelin.es/#keep-it-low-at-all-times) in the CSS Guidelines describes the reasoning behind the workaround in more detail. That web page is quite old know, and some of the topics discussed are solved by recent additions to the CSS standard, but it’s still a valuable resource imo.

For context, this workaround is necessary because custom component-specific styles end up *before* Blueprint’s default styles in our CSS bundle. This is something that I’d like to solve properly at some point (probably when also looking at the rest of the build toolchain), but I haven’t found the time to do so yet.

tillprochaska commented on ui/src/components/Timeline2/TimelineList/TimelineListItem.tsx:
```diff
@@ -0,0 +1,107 @@
+      >
+        <td className="TimelineListItem__date">
+          {start?.value}
+          <br />
```
Not sure, maybe. What’s the downside?

tillprochaska commented:
> Not a massive fan of the Timeline2, but that is just me, and by no means a request for change or a need to refactor later.

This is now gone -- I removed the old implementation and renamed `Timeline2` to just `Timeline`. :)
